(API) Cluster Health report unassigned_primary_shards (#111727)

lukewhiting opened the pull request:
This PR adds a count of currently unassigned primary shards to both the `/_cat/health` and `/_cluster/health` endpoints. This is to aid cluster administrators in estimating the time remaining for a cluster to go from RED to YELLOW status as per enchancement request #111727.

Tests and doc updates are in place with this PR and manual testing with `./gradlew run` has been conducted on the endpoints to ensure correct output.

## Known Limitations
* Testing
  * Due to limitations in the YAML REST test framework skip functionality, YAML REST tests for this endpoint are disabled when running a mixed version cluster by using a cluster version number synthetic feature to skip when any member of the cluster is not at a version greater than when this change is due to be introduced

lukewhiting committed:
WIP #111727 Add new transport version and modify XContent objects

lukewhiting committed:
WIP #111727 Add unassigned primary shards to _cat/health

Also add the unassigned shard count to the /_cat/health endpoint

nielsbauman commented on server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -72,6 +76,11 @@ public ClusterIndexHealth(final IndexMetadata indexMetadata, final IndexRoutingT
             computeRelocatingShards += shardHealth.getRelocatingShards();
             computeInitializingShards += shardHealth.getInitializingShards();
+
+            if (shardHealth.getUnassignedPrimaryShards() > 0) {
+                computeUnassignedPrimaryShards += shardHealth.getUnassignedPrimaryShards();
+            }
```
I don't immediately see why we need this check? If the number of unassigned primary shards is 0, we can still just add it, right?

nielsbauman commented on server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java:
```diff
@@ -197,6 +221,7 @@ public XContentBuilder toXContent(final XContentBuilder builder, final Params pa
         builder.field(ACTIVE_SHARDS, getActiveShards());
         builder.field(RELOCATING_SHARDS, getRelocatingShards());
         builder.field(INITIALIZING_SHARDS, getInitializingShards());
+        builder.field(UNASSIGNED_PRIMARY_SHARDS, getActivePrimaryShards());
```
```suggestion
        builder.field(UNASSIGNED_PRIMARY_SHARDS, getUnassignedPrimaryShards());
```

lukewhiting replied:
Good spot :-) Fixed

nielsbauman commented on server/src/main/java/org/elasticsearch/rest/action/cat/RestHealthAction.java:
```diff
@@ -70,6 +70,8 @@ protected Table getTableWithHeader(final RestRequest request) {
         t.addCell("pri", "alias:p,shards.primary,shardsPrimary;text-align:right;desc:number of primary shards");
         t.addCell("relo", "alias:r,shards.relocating,shardsRelocating;text-align:right;desc:number of relocating nodes");
         t.addCell("init", "alias:i,shards.initializing,shardsInitializing;text-align:right;desc:number of initializing nodes");
+        t.addCell("unassign.primary",
+                  "alias:up,shards.unassigned.primary,shardsUnassignedPrimary;text-align:right;desc:number of unassigned primary shards");
```
I don't think this aligns with our Checkstyle configuration. I don't know if you already got the `Adapter for Eclipse Code Formatter` plugin working already, but in the meanwhile, you can run `gradle spotlessApply` to format all code in compliance with our Checkstyle configuration.

lukewhiting replied:
Have got this all setup now and thanks for the tip on spotlessApply!

lukewhiting committed:
WIP #111727 Add tests for XContent changes

lukewhiting committed:
Swapping round field order

To better match established convention on order of count fields

lukewhiting committed:
Passing API Tests

API YAML tests now pass

lukewhiting committed:
Docs and more tests

Add additional test fields for cluster health API along with doc changes

lukewhiting committed:
Make API tests conditional on API capability

Prevents failure of API checks during cross version testing

lukewhiting committed:
fixup! Make API tests conditional on API capability

Actually remove the field from the backward compatibility tests

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into enhance-111727-active-pri-shards-in-health-endpoint

lukewhiting committed:
Switch to using synthetic features for rest tests

This ensures they are correctly skipped in mixed version clusters where
currently capability checking in the tests does not function correctly

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into enhance-111727-active-pri-shards-in-health-endpoint

gmarouli commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:
```diff
@@ -71,9 +91,96 @@
             $/
+
+---
+"Help - Without unassigned_pri_shard_count":
+  - skip:
+      cluster_features: gte_v8.15.0
+      reason: Capabilities skipping does not yet work in mixed clusters. This will disable the test in such scenarios.
```
Are these tests ever executed? According to the docs:
```
Note that `skip` with `capabilities` or `cluster_features` will skip the test if *any* node in the cluster
has the feature or capability. `requires` will only run the test if *all* of the nodes in the cluster
have the feature or capability.
```
I understand this in the following way:
- Mixed clusters with the current version will be skipped.
- This would only run if the all nodes of the cluster are on version less than 8.15.0

Considering the above I am not sure that these test are worth having. Clusters with mixed versions less than 8.15.0 should have been tested by their own tests. I would like to have mixed cluster tests but I would like one of the nodes to be on the current version.

What do you think?

lukewhiting replied:
I think you are right... With the capabilities API I was intending to test the case of a mixed cluster when you hit an older node as a coordinator but as that API does not work correctly in this case and we had to switch to the features api, I think the test is now redundant.

lukewhiting replied:
Tests removed in latest commit

gmarouli commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java:
```diff
@@ -36,28 +37,7 @@
 @ServerlessScope(Scope.INTERNAL)
 public class RestClusterHealthAction extends BaseRestHandler {
 
-    @Override
-    public List<Route> routes() {
-        return List.of(new Route(GET, "/_cluster/health"), new Route(GET, "/_cluster/health/{index}"));
-    }
-
-    @Override
-    public String getName() {
-        return "cluster_health_action";
-    }
```
I would prefer to keep these two at the top of the class. I am not sure it's convention necessarily, but if you see the majority of the `BaseRestHandlers` have them on top, and it's very helpful to find this metadata at the top of the class.

I understand this is not how we structure classes in general, but in this case it is very helpful considering the nature of the class. Does this make sense to you?

I do not feel so strongly about `allowSystemIndexAccessByDefault`.

gmarouli commented on server/src/test/java/org/elasticsearch/cluster/routing/RoutingTableGenerator.java:
```diff
@@ -129,7 +130,11 @@ public void update(ShardRouting shardRouting) {
             if (shardRouting.initializing()) {
                 initializing++;
             } else {
-                unassigned++;
+                if (shardRouting.primary()) {
+                    unassignedPrimary++;
+                } else {
+                    unassigned++;
+                }
```
If I am not mistaken, here we changed the semantics of the `unassigned`. In the original version of the code it used to count all unassigned shards and now it is counting only the non-primaries. Was that intentional?

nielsbauman commented on server/src/test/java/org/elasticsearch/cluster/health/ClusterIndexHealthTests.java:
```diff
@@ -147,6 +149,7 @@ protected ClusterIndexHealth mutateInstance(ClusterIndexHealth instance) {
             "relocatingShards",
```
This is obviously not your doing, but I'm not a big fan of implementations like these of `mutateInstance`. I prefer implementations like:
https://github.com/elastic/elasticsearch/blob/00f2b9a779adcd52c8172367e6c63132393c8cf9/server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java#L85-L85
because they're much more compact and readable (IMO).

I'll leave it up to you if you want to refactor that in this PR or you keep it like this -- if you even agree, which isn't required of course.

lukewhiting replied:
I agree it's a much better pattern with the vars and the pattern matching instead of a legacy switch.

I might even go one step further and suggest using a builder defined in an anonymous class inside the test. That would further help contain and simplify the mutation method.

That said, I'm going to open a seperate PR for that I think. This one is already quite large and I'm hesitant to introduce more changes.

nielsbauman commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterHealthAction.java:
```diff
@@ -36,28 +37,7 @@
-    @Override
-    public List<Route> routes() {
-        return List.of(new Route(GET, "/_cluster/health"), new Route(GET, "/_cluster/health/{index}"));
-    }
```
Can you revert all the changes here that aren't necessary? I don't see a strong reason to move these things around, so then I'd rather keep them as is. It looks like the only thing you really changed is adding the `supportedCapabilities` method, right?

lukewhiting replied:
Sure. My IDE got a bit enthusiastic with a reformat and applied the default "Static methods go at the top" rule to this class but I agree with you and @gmarouli that in the case of rest actions, it makes sense to have the route at the top.

gmarouli reviewed (changes requested):
Thank you for picking this up @lukewhiting great work! I added a few small comments on the code, let me know what you think.

Furthermore I would like to bring up a bwc concern. 

Currently, in a mixed cluster the count of the unassigned primary shard is not accurate. An older node will not report back the count and the coordinating node will fallback to the value of 0 (which is reasonable). However, the user will get only a partial count of the unassigned primary shards.

I see at least two paths forward:

1. We accept that, the trade-off of fixing this might not be worth it.
2. We use node features for bwc to ensure that when the unassigned primary shard is returned it's always accurate.

What do you think?

nielsbauman commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:
```diff
@@ -1,32 +1,45 @@
-                   pri           .+ \n
-                   relo          .+ \n
-                   init          .+ \n
-                   unassign      .+ \n
```
I won't ask you to revert these changes in this PR, but I personally try to avoid making formatting changes in files like these because there is no enforced format here - and reducing the number of changed lines makes it easier to review (not just the PR, but also the commit in case someone wants to look at this commit in the future).

nielsbauman reviewed (changes requested):
This is great work @lukewhiting! Impressive first PR :)

Most of your changes LGTM, I just left a few small comments.

nielsbauman commented:
@gmarouli
> Currently, in a mixed cluster the count of the unassigned primary shard is not accurate. An older node will not report back the count and the coordinating node will fallback to the value of 0 (which is reasonable). However, the user will get only a partial count of the unassigned primary shards.

@lukewhiting and I also talked about that. I said I was fas in favor of this approach and your option 1 (we accept this the way it is) because this is only relevant in mixed clusters and thus temporarily. My initial thought was that I think it's reasonable that that count isn't accurate when the cluster is in the process of upgrading. On second thought, I guess one could also argue that if there are unassigned primary shards _because_ of the in-progress cluster upgrade, this new count might be valuable. However, I still think it's reasonable to return inaccurate counts during the cluster upgrade.

lukewhiting commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cat.health/10_basic.yml:
```diff
@@ -1,32 +1,45 @@
 ---
-"Help":
+"Help - with unassigned_pri_shard_count":
```
Given your later comment about the other tests in this file, I'm going to revent the name back to the original name as your suggestion and totally remove the "without" tests.

lukewhiting committed:
PR changes

Changes based on feedback from PR review

lukewhiting commented:
@gmarouli 
> Currently, in a mixed cluster the count of the unassigned primary shard is not accurate. An older node will not report back the count and the coordinating node will fallback to the value of 0 (which is reasonable). However, the user will get only a partial count of the unassigned primary shards.

As @nielsbauman mentioned I did initially implemented this so that any not sending the new field through would cause it to be hidden on the rest output however we ended up deciding the additional complexity wasn't worth it for this being a temporary state.

Happy to be convinced otherwise though as I don't really know enough to have a strong opinion yet 😅

gmarouli commented:
> As @nielsbauman mentioned I did initially implemented this so that any node not sending the new field through would cause it to be hidden on the rest output however we ended up deciding the additional complexity wasn't worth it for this being a temporary state.

I guess it depends on the usage of the feature. As @nielsbauman said, if people rely on it then it might be worth it. 

Maybe an idea is to document this. Shall we add a note in the docs that this value will be partially calculated if there nodes in the cluster with versions prior to 8.16, and if a user wants the accurate number they should use the health API?

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into enhance-111727-active-pri-shards-in-health-endpoint

lukewhiting committed:
PR changes - Docs

Add warning about inaccurate count for mixed clusters with node version
below 8.15

lukewhiting commented:
@nielsbauman @gmarouli 

I have added a warning to the docs about inaccurate counts in mixed clusters: https://github.com/elastic/elasticsearch/pull/112024/commits/73e2e06a7319182b2e62894164184f0caab87d34

As we're not 100% sure on the right action here, lets hold off merging and I'll bring it up at the weekly to see what the wider team think.

lukewhiting committed:
Remove typo added during sync with upstream

lukewhiting committed:
Fix off by one error in docs

Corrects the version number the change will be introduced under in the
docs

lukewhiting committed:
Fix for swapped field in test

This was causing unintended mutation

lukewhiting commented:
@elasticsearchmachine buildkite test this

lukewhiting committed:
Fix edge case possible non mutation in test

Current logic made it possible for timeout field to remain unmutated

lukewhiting committed:
Fix another edge case possible non mutation

Current logic made it possible for primaryActive field to remain
unmutated

lukewhiting commented:
Consensus from the team weekly was that the field being incorrect during cluster upgrade should not be an issue as best practice recommendation is that the master node is upgraded last and this endpoint runs on the master node.

With that in mind, the DM team are happy to proceed with the current logic.

gmarouli commented on docs/reference/cluster/health.asciidoc:
```diff
@@ -131,23 +131,26 @@ include::{es-ref-dir}/rest-api/common-parms.asciidoc[tag=cluster-health-status]
     (integer) The number of shards that are not allocated.
 
+`unassigned_primary_shards`::
+    (integer) The number of shards that are primary but not allocated. **Note**: This number may be lower than the true value if your cluster contains nodes running a version below v8.16. For a more accurate count in this scenario, please use the <<cluster-health,cluster health API>>.
```
A small suggestion to be inline with the language in the docs. After a quick look I saw that we do not use the prefix `v`, we just use the version number.
 
```suggestion
    (integer) The number of shards that are primary but not allocated. **Note**: This number may be lower than the true value if your cluster contains nodes running a version prior to 8.16. For a more accurate count in this scenario, please use the <<cluster-health,cluster health API>>.
```

lukewhiting replied:
Sure thing. Have pushed that change :-)

gmarouli reviewed (approved):
LGTM, I added a small comment about how to describe the version in the docs. Apart from that you are good to go 🚀 Congrats, first PR 🥳 !

lukewhiting committed:
Update docs changes to match existing nomenclature

Currently, docs don't use the `v` prefix for version numbers

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into enhance-111727-active-pri-shards-in-health-endpoint

nielsbauman reviewed (approved):
LGTM! Great work on your first PR!! :rocket:

lukewhiting committed:
Bump transport version

Following latest sync with upstream

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

lukewhiting committed:
Update docs/changelog/112024.yaml

elasticsearchmachine commented:
Hi @lukewhiting, I've created a changelog YAML for you.

lukewhiting committed:
Merge branch 'main' of github.com:elastic/elasticsearch into enhance-111727-active-pri-shards-in-health-endpoint

lukewhiting committed:
Merge branch 'enhance-111727-active-pri-shards-in-health-endpoint' of github.com:lukewhiting/elasticsearch into enhance-111727-active-pri-shards-in-health-endpoint
